storage/src/test/java/org/apache/kafka/tiered/storage/actions/ProduceAction.java

```diff
@@ -0,0 +1,163 @@
+        List<BrokerLocalStorage> localStorages = context.getLocalStorages();
+
+        List<LocalTieredStorageCondition> tieredStorageConditions = offloadedSegmentSpecs.stream()
+                .map(spec -> expectEvent(
+                        tieredStorages,
+                        COPY_SEGMENT,
+                        spec.getSourceBrokerId(),
+                        spec.getTopicPartition(),
+                        false))
+                .collect(Collectors.toList());
+
+        // Retrieve the offset of the next record which would be consumed from the topic-partition
+        // before records are produced. This allows consuming only the newly produced records afterwards.
+        long startOffset = context.nextOffset(topicPartition);
+        long beginOffset = context.beginOffset(topicPartition);
+
+        // Records are produced here.
+        context.produce(recordsToProduce, batchSize);
+
+        if (!tieredStorageConditions.isEmpty()) {
+            tieredStorageConditions.stream()
+                    .reduce(LocalTieredStorageCondition::and)
+                    .get()
+                    .waitUntilTrue(OFFLOAD_WAIT_TIMEOUT_SEC, TimeUnit.SECONDS);
+        }
+
+        // At this stage, records were produced, and the expected remote log segments found in the second-tier storage.
+        // Further steps are:
+        // 1) Verify the local (first-tier) storages contain only the expected log segments - that is to say,
+        //    in the special case of these integration tests, only the active segment.
+        // 2) Consume the records and verify they match the produced records.
+        TopicSpec topicSpec = context.topicSpec(topicPartition.topic());
+        long earliestLocalOffset = expectedEarliestLocalOffset != -1L ? expectedEarliestLocalOffset
+                : startOffset + recordsToProduce.size()
+                - (recordsToProduce.size() % topicSpec.getMaxBatchCountPerSegment()) - 1;
+
+        for (BrokerLocalStorage localStorage : localStorages) {
+            // Select brokers which are assigned a replica of the topic-partition
+            boolean isAssignedReplica = context.isAssignedReplica(topicPartition, localStorage.getBrokerId());
+            if (isAssignedReplica) {
+                // Filter out inactive brokers, which may still contain log segments we would expect
+                // to be deleted based on the retention configuration.
+                boolean isActive = context.isActive(localStorage.getBrokerId());
+                if (isActive) {
+                    // Wait until the brokers local storage has been cleared from the inactive log segments.
+                    localStorage.waitForEarliestLocalOffset(topicPartition, earliestLocalOffset);
+                }
+            }
+        }
+
+        List<ConsumerRecord<String, String>> consumedRecords =
+                context.consume(topicPartition, recordsToProduce.size(), startOffset);
```

Comment:
why are we consuming in a produce action? I am assuming that the reason is to validate that produce was executed correctly. yes? Please add a comment about this here.